experimental/storage/layered-map/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+                // descendant will be dropped after we free the `Arc`, which results in a chain
+                // of such structures being dropped recursively and that might trigger a stack
+                // overflow. To prevent that we follow the chain further to disconnect things
+                // beforehand.
+                stack.extend(descendant.drain_children_for_drop());
+            }
+        }
+        self.log_layer("dropped");
+    }
+}
+
+impl<K: Key, V: Value> LayerInner<K, V> {
+    fn new_family(use_case: &'static str) -> Arc<Self> {
+        let family = HashValue::random();
+        Arc::new(Self {
+            root: NodeRef::Empty,
+            children: Mutex::new(Vec::new()),
+            use_case,
+            family,
+            layer: 0,
+            base_layer: 0,
+        })
+    }
+
+    fn spawn(self: &Arc<Self>, child_root: NodeRef<K, V>, base_layer: u64) -> Arc<Self> {
+        let child = Arc::new(Self {
+            root: child_root,
+            children: Mutex::new(Vec::new()),
+            use_case: self.use_case,
+            family: self.family,
+            layer: self.layer + 1,
+            base_layer,
+        });
+        self.children.lock().push(child.clone());
+        child.log_layer("spawn");
+
+        child
+    }
+
+    fn drain_children_for_drop(&self) -> Vec<Arc<Self>> {
+        self.children.lock().drain(..).collect()
+    }
+
+    fn log_layer(&self, event: &'static str) {
+        LAYER.set_with(&[self.use_case, event], self.layer as i64);
+    }
+}
+
+#[derive(Debug)]
+pub struct MapLayer<K: Key, V: Value> {
+    inner: Arc<LayerInner<K, V>>,
+}
+
+/// Manual implementation because `LayerInner` is deliberately not `Clone`.
+impl<K: Key, V: Value> Clone for MapLayer<K, V> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<K: Key, V: Value> MapLayer<K, V> {
+    pub fn new_family(use_case: &'static str) -> Self {
+        Self {
+            inner: LayerInner::new_family(use_case),
+        }
+    }
+
+    pub fn into_layers_view_since(self, bottom_layer: MapLayer<K, V>) -> LayeredMap<K, V> {
+        assert!(bottom_layer.is_family(&self));
+        assert!(bottom_layer.inner.layer >= self.inner.base_layer);
+        assert!(bottom_layer.inner.layer <= self.inner.layer);
+
+        self.log_layer("view");
+        bottom_layer.log_layer("ancestor_ref");
+
+        LayeredMap {
+            bottom_layer,
+            top_layer: self,
+        }
+    }
+
+    pub fn view_layers_since(&self, bottom_layer: &MapLayer<K, V>) -> LayeredMap<K, V> {
+        self.clone().into_layers_view_since(bottom_layer.clone())
+    }
+
+    pub fn log_layer(&self, name: &'static str) {
+        self.inner.log_layer(name)
+    }
+
+    fn is_family(&self, other: &Self) -> bool {
+        self.inner.family == other.inner.family
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct LayeredMap<K: Key, V: Value> {
+    bottom_layer: MapLayer<K, V>,
```

Comment:
why do we need the bottom_layer? could we assume the LRU cache contains all the base key, value pairs?

Reply:
Think "FrozenSparseMerkleTree" -- you need to make sure nodes created between bottom layer and top layer don't go away while you read, by holding a strong ref to the bottom layer root.